### ISSUE TITLE
Improving config resolution performance

### ISSUE THREAD

**Issue by pmdartus** [prettier/prettier#15653]
❤️ First of all, I would to say a big **THANK YOU** to all the maintainers of this great project! I use Prettier on a daily basis, and I have to say that greatly improved my DX. ❤️

After seeing the [interest on Twitter](https://x.com/Vjeux/status/1722733472522142022?s=20) (sorry on X...) for working on a faster alternative to Prettier written in Rust. I was wondering how much could we potentially improve the current Prettier implementation without having to do a full rewrite in Rust.

## Observation

For this purpose, I decided to use the [mui/material-ui](https://github.com/mui/material-ui) project to benchmark Prettier. This project contains ~42K files and roughly 8500 files files that are currently formatted by Prettier. On my machine, it takes 84s to format this project with the [latest](https://github.com/prettier/prettier/commit/3478719cc09e434f2621720389c38ea17a8a61ff) version of the main branch (median over 5 invocations, caching disabled, OSX, Node v20.6.1).

After further investigation, it appears that a sizable amount of time is spent in [`getOptionsForFile`](https://github.com/prettier/prettier/blob/3478719cc09e434f2621720389c38ea17a8a61ff/src/cli/format.js#L336-L339). To validate this observation, I commented the `getOptionsForFile` call and inlined the project configuration. Formatting the entire material-ui project now takes 46s, which is a **45% improvement**.

The fact that it takes 45% of the processing time to resolve file formatting options is highly suspicious. There are certainly some performance improvement opportunities here.

## Digging deeper

As far as I understand, the `getOptionsForFile` essentially resolves and loads the prettier and editor config files in a given directory. Looking at this code it appears to me that Prettier could get a great performance boost from caching the configuration data between getOptionsForFile invocations.

To put this in perspective, I logged all the file system invocations when formatting the material-ui project, the option resolution logic issues: 

* `fs.promises.access`: 827K invocations
* `fs.existsSync`: 100K invocations
* `fs.readFile`: 55K invocations
* `fs.promises.readFile`: 13K invocations

The option resolution logic is delegated to 2 libraries: lilconfig and editorconfig. Both libraries do essentially the same thing, traversing up the file system and looking for specific configuration files. Currently, the configuration lookup isn’t cached between invocations resulting in repeated file system calls.

The main issue here is that it is currently difficult for Prettier to cache the config resolution result between invocations:

* Prettier recently moved cosmiconfig from lilconfig. Unlike the previous config resolution library, lilconfig doesn’t have [cache](https://github.com/antonk52/lilconfig#difference-to-cosmiconfig) support and it appears to be a non-goal for this project. 
* Prettier currently pins the editorconfig dependency to version 0.15.3 but caching support was only introduced in version  [1.0.0](https://github.com/editorconfig/editorconfig-core-js/releases/tag/v1.0.0). As far as I understand, bumping the version isn’t an option because Prettier can’t bundle WASM modules. 

Misc observations:
* [`findProjectRoot`](https://github.com/prettier/prettier/blob/8158a8948a9c94aded9c989660e5c5337ff46ca0/src/config/find-project-root.js) function also recursively traverse the file system upward looking for `.git` or `.hg` directories. Caching the result of its invocation would certainly yield good performance improvement. It would also be ideal to replace the synchronous file system lookup with an async one to avoid blocking the main thread, but I don’t expect major performance improvements as all the files are currently processed serially.
* I think #15363 should be revived. As mentioned in the PR, invoking `.search` on the file itself results in unnecessary file system lookups. It should be possible to optimize this without introducing a breaking change by checking first if the filename references a directory or not. 

## Moving forward

I would be happy to make adequate changes, but I would love to get the maintainers’ feedback first before I invest time into this.

1. I don’t understand the reason for migrating from cosmiconfig to lilconfig ([PR](https://github.com/prettier/prettier/pull/15232)). Is it an option to move back to cosmiconfig to leverage its [caching capabilities](https://github.com/cosmiconfig/cosmiconfig#caching)? If not, I would then open an issue on lilconfig to determine if the maintainer is interested in adding caching support.
2. I don’t see a straightforward way to add support for WASM to the build in order to bump the editorconfig dependency. We can either ask if the maintainer is willing to make the parser swappable to work around this. An alternative approach would be re-implement .editorconfig parsing logic in Prettier. The processing logic looks small enough not to add too much tech debt.  Any thoughts on this?

**Comment by fisker** [comment]
Thanks for the investigation, and pointing out the problems.


> I don’t understand the reason for migrating from cosmiconfig to lilconfig

To have more control of the loaders, and reduce package size.

> I would then open an issue on lilconfig to determine if the maintainer is interested in adding caching support.

It will be nice to have cache feature in lilconfig,  if it doesn't accept, we can also implement ourself.

**Comment by kurtextrem** [comment]
There's definitely a lot of perf gains to have by caching, see https://twitter.com/fabiospampinato/status/1724164039624835073 by @fabiospampinato (just making everyone involved aware of other people's work so that workload can possibly be shared)

**Comment by fabiospampinato** [comment]
The time spent resolving configurations for each file is indeed significant. I think there are mainly 2 issues with that:

1. Things are cached by file path, but for example in MUI's case you can have 10_000 files in the same folder, those files are going to be configured by the same configuration files, because you can't put 10_000 different configuration files for the same thing in the same folder. Basically today the time spent resolving configuration files scales with the number of files times their depth from the your project's root folder. When instead it should scale with the number of folders in the project basically.
2. The file path can still have an impact on the ultimate configuration, since for example `.editorconfig` files can specify different settings for files matching different globs, which shouldn't be too much of a problem because if configuration files are resolved and parsed efficiently that should be the expensive part, and if needed we can start caching that too, but for example looking at the [`editorconfig`](https://github.com/editorconfig/editorconfig-core-js) package it doesn't look like there's any API provided for merging together already pre-parsed configuration files. There's a `parseFromFiles` function, but it's deprecated and it doesn't accept pre-parsed configurations, so that has to get rewritten a bit I think.

FWIW I'm looking into making a more optimized CLI for Prettier, there are these and some other problems to address, just getting a stronger cache in is huge for performance. In roughly 10 days time I should have something to show that can perhaps start to replace the current CLI in the maintainers want it to.

**Comment by antonk52** [comment]
Hi, lilconfig maintainer here. I've added a [PR](https://github.com/antonk52/lilconfig/pull/43) for the cache feature in lilconfig that I am planning to land and release this weekend. There was a [feature request](https://github.com/antonk52/lilconfig/issues/41) from @fisker to add another option to skip reading file content which I am will be adding unless you want to go with a custom config explorer. I can see there is a draft PR linked to this issue.

**Comment by fisker** [comment]
We may want to switch to our own implementation.

**Comment by fisker** [comment]
@pmdartus Would you like to help me test #15663?

**Comment by pmdartus** [comment]
Thanks for putting this PR together @fisker, I took some time to work on this. If you want to migrate away from `lilconfig`, I think we can further optimize and simplify prettier config loading logic: https://github.com/fisker/prettier/compare/config-explorer...pmdartus:pmdartus/config-refactor. 

The most notable change here is instead of creating an explorer object and passing the cache configuration ahead of time, I would recommend passing the cache configuration to `search` and `load` directly. This greatly simplifies the logic by avoiding to maintain multiple searchers.

I ran the same test as before, running prettier against the material UI repo, on the tip of my branch (https://github.com/pmdartus/prettier/commit/9c67d6fb56a8a0bfe56d04168ab9ecde37a1add9) and  I was able to measure a 26% performance improvement (before 84s, after 62s). Just to validate this, I also collected some information about the file system and I was able to measure an 81% reduction in the number of file system lookups.

Even with this change merged, there is still room for improvement to improve config loading. What do you think about keeping this issue opened for now?

**Comment by fisker** [comment]
> The most notable change here is instead of creating an explorer object and passing the cache configuration ahead of time, I would recommend passing the cache configuration to search and load directly.

Yes, ideally it should be that simple, and this is my first implementation(squashed, the history is gone). Later I realize that we used `stopDir` in tests..  https://github.com/prettier/prettier/blob/e31d2f92dd0ba7fc42d2a3eccc74fc705beafbf9/tests/integration/cli-worker.js#L72C7-L72C14

We also considering to accept #8303, so we need that `stopDir`(`stopDirectory` in #15663) option.


>  What do you think about keeping this issue opened for now?

Yes, we should keep it open, at least before we fix `findProjectRoot` and editorconfig resolution.

**Comment by pmdartus** [comment]
> Later I realize that we used stopDir in tests..

Regarding the `stopDir`, it can be passed as an argument to the function to the `search` function as well ([diff](https://github.com/fisker/prettier/compare/config-explorer...pmdartus:pmdartus/config-refactor#diff-bf0c395d1ee7c28ed0b7fda891e4c950b19e48a2403475953abd621c2dfa4b01R62)). This approach also aligns with https://github.com/prettier/prettier/issues/8303. By the way, all the tests are passing on my branch as well, as I took care of updating the mocks in [the tests](https://github.com/fisker/prettier/compare/config-explorer...pmdartus:pmdartus/config-refactor#diff-b8b5b41800db2dde5b0470573f5be6c1975bae76d1c235f2e26b6f1dcbae3eb1R69-R70) to take this into account.

Let me know if you are interested in those changes, and I will open a PR against your branch.

**Comment by fisker** [comment]
Ah, sorry, misread your comment.  At first, I thought we don't need `stopDirectory` support, which make things much simplier.

You change looks good, you can send it

**Comment by pmdartus** [comment]
@fisker PR opened https://github.com/fisker/prettier/pull/1880.

**Comment by fisker** [comment]
Thank you!

**Comment by fisker** [comment]
Let's consider this is done.
